source/h2_connection.c

```diff
@@ -341,25 +328,16 @@ static struct aws_h2_connection *s_connection_new(
             ERROR, connection, "Encoder init error %d (%s)", aws_last_error(), aws_error_name(aws_last_error()));
         goto error;
     }
-
-    /* Initial settings with initial window size and disable push promise for the client */
-    struct aws_http2_setting initial_settings[2];
-    size_t iter = 0;
-    if (!server) {
-        initial_settings[iter].id = AWS_HTTP2_SETTINGS_ENABLE_PUSH;
-        initial_settings[iter++].value = 0;
-    }
-    if (initial_window_size != aws_h2_settings_initial[AWS_HTTP2_SETTINGS_INITIAL_WINDOW_SIZE]) {
-        initial_settings[iter].id = AWS_HTTP2_SETTINGS_INITIAL_WINDOW_SIZE;
-        initial_settings[iter++].value = (uint32_t)initial_window_size;
-    }
     struct aws_h2_pending_settings *init_pending_settings =
-        s_new_pending_settings(connection->base.alloc, initial_settings, iter, NULL, NULL);
+        s_new_pending_settings(connection->base.alloc, &http2_options->initial_settings);
     if (!init_pending_settings) {
         goto error;
     }
-    struct aws_h2_frame *init_settings_frame =
-        aws_h2_frame_new_settings(connection->base.alloc, initial_settings, iter, false /*ACK*/);
+    struct aws_h2_frame *init_settings_frame = aws_h2_frame_new_settings(
+        connection->base.alloc,
+        http2_options->initial_settings.settings_array,
+        http2_options->initial_settings.num_settings,
+        false /*ACK*/);
```

Comment:
Ugh I just realized something nasty:

We're cleaning up `thread_data.pending_settings_queue` (and also `thread_data.pending_ping_queue`) from `s_handler_destroy()`. That doesn't run until all the refcounts on the connection go away. That's why things like `thread_data.active_streams_map` and `synced_data.pending_stream_list` are cleared from `s_finish_shutdown()`. It lets their completion callbacks fire when the connection class is still "valid". If we're not firing all the completion callbacks from `s_handler_destroy()`, then the bindings probably won't release all their refcounts.

But here we are in `s_connection_new()` adding stuff to the `outgoing_frames_queue` and `pending_settings_queue`. If the connection fails to ever set up, then `s_handler_destroy()` never runs.

Two options I can think of:
1) Don't put anything in these lists until s_handler_installed() is called.
2) Clear `pending_settings_queue` lists in both `s_finish_shutdown()` and `s_handler_destroy()`, but don't fire any callbacks from `s_handler_destroy()`

Reply:
Yeah, you are right, I never actually think through how the connection shuts down. The difference between shutdown and destroy. Thank you for pointing it out! Taking the second option here. Since the first one will need h2_connection to keep the initial settings somewhere.

Reply:
Taking the first option instead, since the user data is not valid until the handler_installed...